File: Data Structures & Algorithms/count-number-of-islands/submission-2.kt
class Solution {
    fun numIslands(grid: Array<CharArray>): Int {
        val m = grid.size
        val n = grid[0].size
        fun dfs(i: Int, j: Int) {
            if (i in 0..m-1 && j in 0..n-1 && grid[i][j]== '1')
            {
                grid[i][j] = '0'
                dfs(i-1,j)
                dfs(i,j-1)
                dfs(i+1,j)
                dfs(i,j+1)
            }
        }
        var res= 0
        for (i in 0..m-1)
            for (j in 0 .. n-1)
                if (grid[i][j] == '1') {
                    dfs(i,j)
                    res++
                }
                
        return res
    }
}

File: Data Structures & Algorithms/max-area-of-island/submission-2.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m,n = len(grid), len(grid[0])
        def dfs(i,j):
            if not 0<= i < m or not 0 <= j < n or grid[i][j] != 1:
                return 0

            grid[i][j] = 0
            res = 1+ dfs(i,j+1) + dfs(i,j-1) + dfs(i-1,j)+ dfs(i+1,j)

            return res 
        res = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    count = dfs(i,j)
                    res = max(res, count)
        return res